src/Storage/Hydration/EventHydrator.php

```diff
@@ -0,0 +1,167 @@
+use ReflectionProperty;
+use function array_merge;
+use function sprintf;
+use function ucfirst;
+
+class EventHydrator
+{
+    private $classResolver;
+
+
+    public function __construct(EventClassResolver $classResolver)
+    {
+        $this->classResolver = $classResolver;
+    }
+
+
+    /**
+     * @param mixed[] $data
+     */
+    public function hydrate(array $data) : Event
+    {
+        if (!isset($data['eventType'])) {
+            throw new HydrationFailed('Cannot hydrate event without an event type.');
+        }
+
+        $class = $this->classResolver->resolve($data['eventType']);
+
+        try {
+            $reflection = new ReflectionClass($class);
+        } catch (ReflectionException $exception) {
+            throw new HydrationFailed(
+                sprintf("Cannot hydrate event, class '%s' could not be found.", $class),
+                0,
+                $exception
+            );
+        }
+
+        $event      = $reflection->newInstanceWithoutConstructor();
+        $properties = $this->getProperties($reflection);
+
+        if (!($event instanceof Event)) {
+            throw new HydrationFailed(
+                sprintf("Cannot hydrate event as the class '%s' is not an instance of event.", $class)
+            );
+
+        }
+
+        foreach ($properties as $property) {
+            $this->setEventProperty($event, $property, $data);
+        }
+
+        return $event;
+    }
+
+
+
+    /**
+     * @return mixed[]
+     */
+    public function dehydrate(Event $event) : array
+    {
+        $data = [];
+
+        $reflection = new ReflectionClass($event);
+        $properties = $this->getProperties($reflection);
+
+        foreach ($properties as $property) {
+            $data[$property->getName()] = $this->getEventPropertyValue($event, $property);
+        }
+
+        $data['eventType'] = $event::eventType();
+
+        return $data;
+    }
+
+
+    /**
+     * @return mixed
+     */
+    private function getEventPropertyValue(Event $event, ReflectionProperty $property)
+    {
+        // First check to see if we have a getter method. If we do, use that to get the value instead.
+        $method = $this->findGetterForProperty($property);
+
+        if ($method) {
+            return $method->invoke($event);
+        }
+
+        $property->setAccessible(true);
+
+        return $property->getValue($event);
+    }
+
+
+    /**
+     * @param mixed[] $data
+     */
+    private function setEventProperty(Event $event, ReflectionProperty $property, array $data) : void
+    {
+        if (!isset($data[$property->getName()])) {
+            return;
+        }
+
+        $property->setAccessible(true);
+        $property->setValue($event, $data[$property->getName()]);
+    }
+
+
+    private function findGetterForProperty(ReflectionProperty $property) : ?ReflectionMethod
+    {
+        $reflection = $property->getDeclaringClass();
+        $methodName = sprintf('get%s', ucfirst($property->getName()));
```

Comment:
Looking for a method with the same name as the property might be useful too. Also for boolean properties, supporting `is` / `has` etc would be nice?

Reply:
It does look for a method with the same property method if this doesn't yield a result below.